functable.c

```diff
@@ -8,450 +8,497 @@
 #include "crc32_braid_p.h"
 #include "deflate.h"
 #include "deflate_p.h"
-
 #include "functable.h"
-
 #include "cpu_features.h"
 
-Z_INTERNAL Z_TLS struct functable_s functable;
+static void init_functable(void) {
+    uint32_t (* adler32)            (uint32_t adler, const uint8_t *buf, uint64_t len);
```

Comment:
Where did global `struct functable_s functable;` go?

Reply:
Copy one struct to another is not atomic on all CPUs.
Copy one pointer to another is atomic on most CPUs.

But I think it's better to use an intermediate ``struct functable_s`` instead of separate variables.

Reply:
> Copy one struct to another is not atomic on all CPUs.
Copy one pointer to another is atomic on most CPUs.

It should have a comment where it copies each one individually telling why.

Reply:
Yes, functable_s is a better idea. Copying whole table vs pointer by pointer wouldn't make a difference as long as individual pointers are copied atomically.

Reply:
> Copying whole table vs pointer by pointer wouldn't make a difference as long as individual pointers are copied atomically.

**No.** There are no guarantees how struct copying will be implemented.
The compiler can use ``rep movsb`` (x86) and then the copy will be byte by byte, not pointer to pointer (by pointer-size copy instruction).

Reply:
@pps83 
I have a question, can the C compiler replace element-by-element struct copying with memcpy?
If yes, use struct are dangerous.

Reply:
I think that it can be better to copy pointers as follows:

``__atomic_store`` (gcc-4.7+, clang, other gcc-compatible, https://gcc.gnu.org/onlinedocs/gcc-4.7.0/gcc/_005f_005fatomic-Builtins.html)
Performance note:

https://github.com/boostorg/atomic/blob/c277de6b49804f759ccc429ce56fdc03684050b0/include/boost/atomic/detail/core_ops_gcc_atomic.hpp#L67-L78

or
``InterlockedExchangePointer`` (MSVC, https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedexchangepointer)

Reply:
@phprus interlocked exchange could be used if TLS is removed. In this PR it's not needed because of TLS.

However, it seems that TLS is used because zlib-ng assumes that multiple cores on a CPU might have different features, or, to put it in other words, zlib-ng assumes that CPU tests have to be done for each thread. Not sure if that's a valid or useful assumption, simply because after `functions` table was initialized for the CPU/thread it doesn't mean that the thread will continue running on the same core.

Reply:
In some architectures a thread is pinned to specific core, so it's safe to assume that the supported features don't change after initial check when using TLS. This applies to C code only, as with C++ the thread can switch cores if exception is thrown. 

Without TLS, there is also very small window during initialisation of function table, where another thread possibly could call same function and get different result or even invalid pointer. This is why we didn't initialise all the function table members in every stub function as that reduced the window to maximum of few nanoseconds.

ARM is one example of architecture that is known to mix cores with different supported features.

Reply:
@mtl1979
C and C++ threads can be run by the OS-scheduler on any available hardware core.
To bind a thread to the hardware core, we must explicitly set the thread affinity. TLS binds data to the program thread, not to the CPU core.
Therefore, using TLS does not protect against CPUs with different feature sets.

Reply:
zlib or zlib-ng doesn't start threads itself, so it doesn't have to worry about setting the affinity either... Even in my own programs, I haven't ever needed to set the thread affinity myself, it does that automatically.

If program is supposed to run on processor with mismatched cores, it would likely crash even without using zlib or gzip functions if it doesn't set thread affinity either implicitly or explicitly.

Reply:
That’s what I also thought, that something like windows’  SetThreadAfinity needs to be used to force thread to be executed on specific core.

Arm cpus had low/high power cores for a long time, new x86 CPUs also now have the same stuff. Do these low/high power cores have different CPU features?? I’d expect them to be the same, otherwise this would complicate lots of stuff.

Reply:
@pps83 
@mtl1979 

So, if TLS remains anyway...

In this case, there will be no conflicts when using ``functable``.
We can simplify PR by removed the temporary local variable ``struct functable_s ft`` and do the assignment directly to the members of ``functable``.

Using a common ``init_functable`` function will speed up all calls to versioned functions except the only first call of any versioned function in each thread.
The slowdown of the first call will be less than the total slowdown of all functions before RP, since the compiler will most likely combine the same if.

Reply:
@mtl1979


Single ``init_function`` will help solve another issue.

All CPU feature flags is not TLS:
https://github.com/zlib-ng/zlib-ng/blob/41a5e00e7c9fd8fce5c2135fe412d4e0f3f15eb3/arch/arm/arm_features.c#L74-L75

https://github.com/zlib-ng/zlib-ng/blob/41a5e00e7c9fd8fce5c2135fe412d4e0f3f15eb3/arch/x86/x86_features.c#L21-L32

This makes useless to use TLS for ``functlable``.

After this PR (single ``cpu_check_features()`` call) we replace ``cpu_check_features()`` and ``*_cpu_has_*`` variables with this code:

```c

struct x86_cpu_features
{
    bool x86_cpu_has_...;
};

struct arm_cpu_features
{
    bool arm_cpu_has_...;
};

void x86_check_features(struct x86_cpu_features *f);
void arm_check_features(struct arm_cpu_features *f);

// Plafrom depended:
#define cpu_features       x86_cpu_features;
#define cpu_check_features x86_check_features


static void init_functable(void) {

...
cpu_features cpu;
cpu_check_features(&cpu);

...
if (cpu.*_cpu_has_*) {
    ...
}

...

```

This logic ensures that if an external application binds threads to the CPU correctly, then zlib-ng will always choose the correct implementation for the thread/CPU.

Reply:
@phprus It's true that currently checking the CPU features doesn't utilize thread-local storage or is thread-safe... None of us long-time contributors is using mismatched SMP setup as their development machine, so there wasn't easy way to test the change.

As a sidenote, TLS support can still be disabled, so any change can't break thread-safety more than it is already broken.

Reply:
> After this PR (single `cpu_check_features()` call) we replace `cpu_check_features()` and `*_cpu_has_*` variables with this code:
> 
> ```c
> struct x86_cpu_features
> {
>     bool x86_cpu_has_...;
> };
> 
> struct arm_cpu_features
> {
>     bool arm_cpu_has_...;
> };
> 
> void x86_check_features(struct x86_cpu_features *f);
> void arm_check_features(struct arm_cpu_features *f);
> 
> // Platform dependent:
> #define cpu_features       x86_cpu_features;
> #define cpu_check_features x86_check_features
> 
> 
> static void init_functable(void) {
> 
> ...
> cpu_features cpu;
> cpu_check_features(&cpu);
> 
> ...
> if (cpu.*_cpu_has_*) {
>     ...
> }
> 
> ...
> ```
> 
> This logic ensures that if an external application binds threads to the CPU correctly, then zlib-ng will always choose the correct implementation for the thread/CPU.

Some could argue that thread-safety of CPU feature checks should indeed be separate PR, but reviewed before instead of after this PR.

Reply:
@phprus you broke my nuclear weapon :).
I recall seeing TLS, made some discussion years ago (perhaps a pr or an issue on gh), saw that was total nonsense. I did remove that mess from my fork. I wanted to see how far that would go this time. Now all the maintainers could be reasonably sure that ALL that discussions, and code related to TLS use was pretty much nonsense, and the only potentially valid reason was atomic pointer swap. IMO, all the stubs should be removed, `functable` should be zero init, `inflateInit`/`deflateInit` should call `functable_init` (to avoid pointless discussion, by `inflateInit`/`deflateInit` I also mean all the functions that could be first-time called by the user code).

Reply:
> The slowdown of the first call will be less than the total slowdown of all functions before RP, since the compiler will most likely combine the same if.

I think any reasonably programmer would understand that it's pointless to discuss performance in this case. It just make no sense at all, it distracts from stuff that matters. The code had bugs, and mystical perf is second to correctness, especially when it's crystal clear that this thing has no perf overhead. It's not like it called once per 100-byte block. Even if it was called once per compres/uncompress sequence it's still would be a non-issue.

Reply:
I've been programmer for few months short of 40 years.... And I see your argument is still overly simplified... I'm not saying this PR is utter waste of time, I'm just saying it can be further improved before we approve it.

Reply:
@pps83 
I apologize :)

Reply:
@mtl1979

> Some could argue that thread-safety of CPU feature checks should indeed be separate PR, but reviewed before instead of after this PR.

I don't understand you.

If we support CPU with different functions on different cores we need TLS.
Then the current implementation is broken since ``cpu_check_features`` does not use TLS. TLS is not free, so it should be used less.
Because of this, I think it's best to have one point of ``cpu_check_features`` call and select for all versioned features.

If we don't we not need TLS.
In this case we need atomic initialization of pointers in ``functable``. And it will also be easier to do it in one place.

So I guess changing ``cpu_check_features`` will be easier after this PR.

Reply:
I didn't write the current implementation of CPU feature checking... It was written gradually a long time before we even had function table. Like I said elsewhere in this PR, I already know the current implementation is buggy, but instead of introducing more bugs or regressions, we should fix the current ones first...

I've been developer for long enough time to notice when people try to mix opinions with facts. This will only complicate reviewing pull requests as none of us know everything about every part of the code, so when people try to confuse us, we need to research to make sure what information we have is actually factual and not just result of trolling.

A lot of time I might sound like a real bitch or arrogant, but that is just to make sure every pull request gets reviewed properly so we don't need to revert or rewrite them later wasting valuable time. @nmoinvaz and I have been talking about improving the CI to catch even more issues, but those improvements have not been merged in yet.